Refactor Query Expressions (Filter, Order, Pagination)

mcaskill opened the pull request:
An internal rewrite of the aforementioned classes. The biggest change is the addition of a base class called `AbstractExpression` which allows room to expand our "query building".

**Requirements**
- [locomotivemtl/charcoal-property:0.5.1](https://github.com/locomotivemtl/charcoal-property/releases/tag/0.5.1) for `StorablePropertyTrait::fieldNames()`

**Breaking Changes**
- `DatabaseSource`
  - Renamed `DEFAULT_DB_TYPE` to `MYSQL_DRIVER_NAME`
- `Pagination`
  - Changed `DEFAULT_PAGE` from 0 to 1 (interchangeable)
  - Renamed `DEFAULT_NUM_PER_PAGE` constant to `DEFAULT_COUNT`
- `Filter`
  - Moved `Filter::DEFAULT_TABLE_NAME` to `DatabaseSource::DEFAULT_TABLE_ALIAS`
  - Removed `DEFAULT_FUNC`, "func" is now `NULL` by default
  - Renamed `val` to `value`
  - Renamed `string` to `condition`
  - Renamed `table_name` to `table`
- `Order`
  - Renamed `string` to `condition`
  - Renamed `table_name` to `table`

Deprecation warnings for "val", "string", and "table_name" can be ignored with `error_reporting(E_ALL ^ E_DEPRECATED);`.

**Highlights**
- `DatabaseOrder`
  - If the Order's mode is `NULL`, the "custom" mode is used (if a "condition" is defined) and takes precedence over "values" mode (if "values" is defined).
  - Added "direction"
- `DatabaseFilter`
  - With `StorablePropertyTrait::fieldNames()`, a Filter can now predicate multiple columns (using an `OR` operand)
  - If "condition" is defined, it takes precedence over the expression's other features.
- `Filter`, `Order`
  - Moved table name and property definitions to `FieldTrait` and `FieldInterface` to unify property assignments to query expressions
- Added support for named expressions
- Improved support for retrieving a property's field names (e.g., `name_en`, `name_fr`)

**Todo**
- [x] Rename `Expression::$string`  to `Expression::$condition`
- [x] Rename `string` to `condition` to reduce ambiguity of custom expressions
- [x] Rename `table_name` to `table` to reduce ambiguity with a table alias
- [x] Rename `val` to `value` to line-up with `values` from `Order`
- [x] Add `DatabaseExpressionInterface` with `sql()` method
- [x] Change database expressions to implement `DatabaseExpressionInterface`

mcaskill committed:
Cleanup Source Classes

Changes:
- Cleaned up imports
- Added missing imports
- Added breathing room around array values
- Converted string FQCLN with "class" constants

mcaskill committed:
Refactor Filter, Order, Pagination

Notes:
- Pagination: Changed DEFAULT_PAGE from 0 to 1
- Pagination: Renamed DEFAULT_NUM_PER_PAGE constant to DEFAULT_COUNT
- Filter: Moved DEFAULT_TABLE_NAME to `DatabaseSource::DEFAULT_TABLE_ALIAS`
- Filter: Removed DEFAULT_FUNC, "func" is now NULL by default
- Field names (property idents) are now quoted for the SQL (see `FieldTrait::quoteIdentifier()`)

Changes:
- Pagination converts page number 0 to 1
- Added AbstractExpression and ExpressionInterface as basis for Filter, Order, and Pagination
- Added "name" to identify expressions
- Added Field Mixin to share features between Filter and Order
- Updated classes and unit tests to reflect changes to Filter, Order, Pagination
- Added unit tests for AbstractExpression
- Refactored unit tests for Filter, Order, Pagination

mcaskill committed:
Revise FieldTrait

Changes:
- Moved `FieldTrait::quoteIdentifier()` to `AbsractExpression` as static method
- Added unit tests for `FieldTrait`
- Added unit tests for `quoteIdentifier()`
- Added multilingual support to `ContainerProvider`

mcaskill committed:
Bump branch-alias to 0.3

mcaskill committed:
Refactor database expressions and its unit tests

mcaskill commented:
See #6 for an implementation of "filter groups".

---

With a base class (and `ExpressionInterface`), we could more easily implement "filter groups" as currently proposed in #4.

What I propose is:
- Add `Group` which extends `AbstractExpression`
- Add `Group::$conditions` which represents a tree-like array structure of strings or other expression objects to be added or nested inside this object.
- Move `Filter::$operand` to `Group::$conjunction` o be the glue that will join all the conditions at this level of the expression tree.
- Modify `DatabaseSource::sqlFilters()` to nest all added filters under one master `Group` when compiling to a string to contain a correctly parenthesized and nested expression.
- Modify `Filter` and `Order` to extend `Group`

```
"filters": [
	{
		"property": "active",
		"val": true
	},
	{
		"conjunction": "OR",
		"conditions": [
			{
				"property": "categories",
				"val": "foobar"
			},
			{
				"property": "{{ recreationPropertyKey }}",
				"val": "{{ id }}"
			}
		]
	}
]
```

mducharme commented on tests/Charcoal/ContainerProvider.php:
```diff
@@ -131,6 +134,64 @@ public function registerTranslator(Container $container)
+     * @param  Container $container A DI container.
+     * @return void
+     */
+    public function registerMultilingualTranslator(Container $container)
```
Please remove from this PR.

mcaskill commented:
Todo:
- [ ] Rename `AbstractExpression::$string`  to `AbstractExpression::$raw`
- [ ] Add `DatabaseExpressionInterface` with `sql()` method
- [ ] Change database expressions to implement `DatabaseExpressionInterface`
- [ ] Add `Group` which extends `AbstractExpression`
- [ ] Change `Filter` and `Order` to extend `Group`

mcaskill committed:
Add DatabaseExpressionInterface

mcaskill committed:
Improve Source Classes

Changes `AbstractSource`
- Simplified `$pagination` usage
- Simplified case-transformation for `getter()` and `setter()`
- Added `resolvePropertyName()` for extracting a property identifier
- Added support for optionally passing data to `createFilter()`, `createOrder()`, and `createPagination()`
- Added `hasPagination()` method

Changes `DatabaseSource`:
- Replaced class constant `DEFAULT_DB_TYPE` with `MYSQL_DRIVER_NAME`
- Added class constant `SQLITE_DRIVER_NAME`
- Use driver name constants where needed
- Added `hasTable()` method
- Added `sqlFrom()` method
- Use `Expression::quoteIdentifier()`
- Added support for converting basic expressions into their DatabaseExpression equivalent
- Fixed `getModelFields()` when given a subset of properties; assure the list is unique to prevent unnecessary reprocessing

Changes:
- Documented class properties and methods
- Revised methods and internal variables
- Reorganized location of certain methods

mcaskill committed:
Rename expression option "val" to "value"

Changes:
- Replaced `val()` and `setVal()` with `value()` and `setValue()` respectively
- Added deprecation notice when "val" is used in `Expression::setData()` and reassign the value to "value"

mcaskill committed:
Refactor Filter, Order, Pagination, Expression

Changes:
- Moved "condition" from `AbstractExpression` to `Expession` (and `RawExpressionInterface`)
- Rename `Field` mixin to `ExpressionField`
- Modified behaviour of `data()` method to return all options
- Modified behaviour of `jsonSerialize()` method to return only mutated options
- Renamed methods in database expression classes
- Revised documentation of expression classes, interfaces, and methods
- Reorganized and revised unit tests

mcaskill committed:
Rename expression option "string" to "table_name"

Changes:
- Replaced `string()`, `setString()`, and `hasString()` with `condition()`, `setCondition()`, and `hasCondition()` respectively
- Added deprecation notice when "string" is used in `AbstractExpression::setData()` and reassign the value to "condition"
- Replaced `tableName()`, `setTableName()`, and `hasTableName()` with `table()`, `setTable()`, and `hasTable()` respectively
- Added deprecation notice when "table_name" is used in `setData()` and reassign the value to "table"

mcaskill committed:
Add "direction" to Order expression

Changes:
- Added description of SQL resolution for `DatabaseOrder`
- Added "direction" option for controlling the sorting of properties and values
- Assigning a direction to `Order::setMode()` affects "direction"
- Mutating "values" via `Order::setData()` affects "mode"
- Updated unit tests

mcaskill committed:
Improve Filter expression

Changes:
- Added description of SQL resolution for `DatabaseFilter`
- Added "value" validation  for certain predications
- Added new operators to `Filter`
- Improved `AbstractExpression::parseValue()`
- Updated and improved unit tests

mcaskill committed:
Add FilterCollection + OrderCollection Mixins

Changes:
- Moved Filter and Order collecting to their own traits and interfaces
- Modified AbstractSource to feature a backwards-compatible method signature for adding Filters and Orders
- Moved Model-based post-processing of added Filters to `AbstractSource::parseFilterWithModel()`
- Moved Model-based post-processing of added Order to `AbstractSource::parseOrderWithModel()`
- Cleaned up CollectionLoader methods and documentation
- Added unit testing for Filter and Order collection mixins

mcaskill committed:
Improve AbstractSource unit testing

Changes:
- Reorganized unit testing utilities into new `AssertionsTrait` and `ReflectionsTrait`
- Completed unit testing for `AbstractSource`

mcaskill committed:
Improve StorableTrait, AbstractModel

Changes `StorableTrait`:
- Improved unit testing
- Improved logging of failed saves / updates / deletes
- Documented `$logger` class property in class docblock
- Updated `SourceMock` and `StorableMock` to reflect current state of classes

Changes `AbstractModel`:
- Removed extraneous `AbstractModel::save()`
